template/registry/registry.go

```diff
@@ -0,0 +1,146 @@
+package registry
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+)
+
+type Registry map[string]Func
+
+// Func is a function that can be used in formatting and matching contexts.
+type Func struct {
+	FormatSignature *Signature                             // Signature of the function when called in formatting context
+	MatchSignature  *Signature                             // Signature of the function when called in matching context
+	Fn              func(any, map[string]any) (any, error) // Function itself
+	Name            string
+	Description     string
+}
+
+// Signature is a signature of the function, i.e. what input and options are allowed.
+type Signature struct {
+	ValidateInput   func(any) error
+	Options         Options
+	IsInputRequired bool
+}
+
+// Option is a possible options for the function.
+type Option struct {
+	Name        string
+	Description string
+	Default     any
+
+	ValidateValue  func(any) error // If option value is not restricted by a set of values.
+	PossibleValues []any           // If option value is restricted by a set of values.
+}
+
+type Options []Option
+
+// NewRegistry returns a new registry with default functions.
+func NewRegistry() Registry {
+	return Registry{
+		"string": *stringRegistryF,
+		"number": *numberRegistryF,
+	}
+}
+
+func (f *Func) Format(input any, options map[string]any) (any, error) {
+	if f.FormatSignature == nil {
+		return "", fmt.Errorf("function '%s' is not allowed to use in formatting context", f.Name)
+	}
+
+	if err := f.FormatSignature.check(input, options); err != nil {
+		return "", fmt.Errorf("check input: %w", err)
+	}
+
+	return f.Fn(input, options)
+}
+
+func (f *Func) Match(input any, options map[string]any) (any, error) {
+	if f.MatchSignature == nil {
+		return "", fmt.Errorf("function '%s' is not allowed to use in selector context", f.Name)
+	}
+
+	if err := f.MatchSignature.check(input, options); err != nil {
+		return "", fmt.Errorf("check input: %w", err)
+	}
+
+	return f.Fn(input, options)
+}
+
+func (s *Signature) check(input any, options map[string]any) error {
+	if s.IsInputRequired && input == nil {
+		return errors.New("input is required, got nil")
+	}
+
+	if !s.IsInputRequired && input != nil {
+		return errors.New("input is not allowed")
+	}
+
+	if len(s.Options) == 0 && len(options) > 0 {
+		return errors.New("options are not allowed")
+	}
+
+	if s.ValidateInput != nil {
+		if err := s.ValidateInput(input); err != nil {
+			return fmt.Errorf("validate input: %w", err)
+		}
+	}
+
+	if len(options) > 0 {
+		if err := s.Options.check(options); err != nil {
+			return fmt.Errorf("check options: %w", err)
+		}
+	}
+
```

Comment:
Let the `check()` handle empty options: keeps the flow nice and tidy.

```suggestion
	if err := s.Options.check(options); err != nil {
		return fmt.Errorf("check options: %w", err)
	}
```

Reply:
`check` already handles empty options nicely